### PR TITLE
docgen: Optimize README update script

### PR DESCRIPTION
**Pull request by aduth**
Previously: #15679, #15200

This pull request seeks to optimize and refactor the `bin/update-readmes.js` script. The resulting changes should decrease runtime of this script by **90% or more** for typical usage (baseline of 6.65s to 0.6s `lint-staged` single package or 1.275s complete run average).

The goal here was largely to reduce the delay in `pre-commit` tasks.

**Implementation Notes:**

Influenced by previous optimizations to the Gutenberg build script in #15230, the approach here uses a [`fast-glob`](https://www.npmjs.com/package/fast-glob) stream, allowing compilation to begin even before all files are known.

Like in #15200, the process is now again asynchronous, but incorporates necessary revisions from #15679 to ensure that tokens within a single file can only be replaced in sequence, not in parallel, to avoid conflict. 

It avoids a hard-coded list in favor of reading in the contents of discovered README files to determine whether replacement tokens exist. A "new" syntax was incorporated for files which contain multiple tokens that source from separate files (see "Autogenerated actions", "Autogenerated selectors"). This is really more a convention considered by `update-readmes.js`, and doesn't require any revisions to the `docgen` tool.

Finally, since `lint-staged` will already [provide the staged files as arguments to the script](https://github.com/okonet/lint-staged#why), we can use this to filter packages to those where modifications have occurred.

**Testing Instructions:**

Verify that changes to a file covered by `docgen` will update the corresponding README, including those with multiple or custom "Autogenerated" tokens (e.g. `core-data`).

Verify that the behavior of `lint-staged` is not regressed. You can do this by editing a JavaScript file, staging it (`git add`), then running `npx lint-staged`.

### CODE REVIEW

**Comment by nosolosw** [comment]
The performance gains that something like this could yield are great! Thanks for working on this.

A couple of things:

- This doesn't work with node latest v10 LTS (it does with v12 LTS). Executing the script directly `node ./bin/update-readmes.js` yields some error messages that can help investigate.
- I'd suggest testing with the worst-case scenario: that both the package's README and the handbook's data API docs need to be updated. Are the changes in `bin/update-readmes.js` portable to `./docs/tool/update-data.js`?

## Testing instructions for the worst-case scenario

```bash
# edit the JSON comment of an exported entity in `packages/core-data/src/actions.js`
git add packages/core-data/src/actions.js
git commit -m 'testing'
```

The expected result is that two files should contain changes to update: `packages/core-data/README.md` and `docs/designers-developers/developers/data/data-core.md`.

File: bin/update-readmes.js [hunk 1]
```diff
@@ -1,70 +1,138 @@
...
+ * @type {string[]}
+ */
+const files = process.argv.slice( 2 );
+
+glob.stream( `${ PACKAGES_DIR }/${ getPackagePattern( files ) }/README.md` )
+	.pipe( filterTokenTransform )
+	.on( 'data', async ( /** @type {WPReadmeFileData} */ data ) => {
+		const [ file, tokens ] = data;
+
+		// Each target operates over the same file, so it needs to be processed
```
Comment (nosolosw) [inline bin/update-readmes.js]: This comment could be improved to something along these lines:

```
		// Each file can have more than one placeholder content to update
		// (which are represented by tokens). docgen updates one token at a time,
		// so we need to fill the tokens sequentially as to prevent that
		// the processes override each other.
```

File: bin/update-readmes.js [hunk 2]
```diff
@@ -1,70 +1,138 @@
...
+		// synchronously, to make sure the processes don't overwrite eachother.
+		for ( const [ token, path ] of tokens ) {
+			const packageName = getFilePackage( file );
```
Comment (nosolosw) [inline bin/update-readmes.js]: This can be taken out of the loop.

File: bin/update-readmes.js [hunk 3]
```diff
@@ -1,70 +1,138 @@
...
+		return '*';
 	}
 
-	const [ packageName, targets ] = entry;
-
-	Object.entries( targets ).forEach( ( [ token, path ] ) => {
-		// Each target operates over the same file, so it needs to be processed synchronously,
-		// as to make sure the processes don't overwrite each other.
-		const { status, stderr } = spawnSync(
-			join( __dirname, '..', 'node_modules', '.bin', 'docgen' ).replace( / /g, '\\ ' ),
```
Comment (aduth) [inline bin/update-readmes.js]: Note: The changes here remove the `String#replace` fix which was introduced by @MarkMarzeotti in #18253 to improve Windows support.

My reading of the documentation of `execa` leads me to believe that this should be handled by default in the library, as well as potential other issues we could have with Windows support:

>Node has issues when using spawn on Windows:
>
>[...]
>- Has problems running commands with [spaces](https://github.com/nodejs/node/issues/7367)

https://github.com/moxystudio/node-cross-spawn#why

File: bin/update-readmes.js (hunk 1 above)
Comment (aduth) [reply bin/update-readmes.js]: > This comment could be improved to something along these lines

Yep, I didn't think to update this, but I agree, terms like "target" don't make much sense anymore. Updated in 41908161d5f78dab12727566caaa2e5c1ebdc474.

File: bin/update-readmes.js (hunk 2 above)
Comment (aduth) [reply bin/update-readmes.js]: > This can be taken out of the loop.

Yes, even more optimization! 😄 Updated in 3958615ab937d7f8d08010ea19c03fc354465513.

**Comment by aduth** [comment]
> * This doesn't work with node latest v10 LTS (it does with v12 LTS). Executing the script directly `node ./bin/update-readmes.js` yields some error messages that can help investigate.

Well, my understanding is that we target the current LTS for support. This was documented previously, but apparently was removed as part of the changes in #17004. I'm not really sure why, because it leaves some ambiguity to what we support (commented at https://github.com/WordPress/gutenberg/pull/17004#discussion_r354005575).

Personally, I think we should restore this documented expectation, and not care to support v10 and earlier.

**Comment by aduth** [comment]
> * I'd suggest testing with the worst-case scenario: that both the package's README and the handbook's data API docs need to be updated. Are the changes in `bin/update-readmes.js` portable to `./docs/tool/update-data.js`?

I wasn't aware of this other script. It seems it's not updated correctly, as per your suspicion. Looking at the code, I'd guess at the very least the keys of "Autogenerated selectors" would need to be updated, else try to adapt similar changes to what was applied here for `update-readmes.js` into `update-data.js`.

**Comment by nosolosw** [comment]
> Personally, I think we should restore this documented expectation, and not care to support v10 and earlier.

I'm fine with this. However, I wanted to note that, at the moment, node [has active two LTS lines](https://nodejs.org/en/about/releases/) and v10 will be LTS until April 2020. Don't know why they do this and I also don't see why we wouldn't upgrade to v12 given that it's easy to install it into any system git a nvm/n kind of setup, but wanted to share anyway to not leave anyone behind.

**Comment by aduth** [comment]
> However, I wanted to note that, at the moment, node [has active two LTS lines](https://nodejs.org/en/about/releases/) and v10 will be LTS until April 2020.

Followed-up at https://github.com/WordPress/gutenberg/pull/18923#issuecomment-562198455.

That's a good observation. In the updated documentation in #18923, I suggest we note this as "latest active LTS", especially when we're already promoting use of `nvm` which will install on this basis anyways.

**Comment by aduth** [comment]
On the topic of `docs/tool/index.js`, I would really like if we could find some way to consolidate these to a single script. I'm still trying to think through how that might look. Do you have any thoughts on this?

I think the use of "Autogenerated actions" and "Autogenerated selectors" could serve as some sort of distinction of what should qualify as data documentation, but there's a few edge cases to consider:

- We include `core-data` action/selector documentation in its README.md
- We effectively rename the `core-data` package to "core" for its data documentation (hard-coded rename)

**Comment by nosolosw** [comment]
@aduth now that https://github.com/WordPress/gutenberg/pull/18820 has landed, I've rebased this in an attempt to ease the burden from you. I hope it's useful, but I still have the old branch locally, in case you rather want it back.

This doesn't work as expected as per the instructions at https://github.com/WordPress/gutenberg/pull/18840#issuecomment-561751133 (readme package docs aren't updated). We can port the changes to `docs/tool/update-data` when it does.

**Comment by aduth** [comment]
@nosolosw Thanks for helping out with the refresh! I haven't had a chance yet to review the updates in much detail, but it looks good at a glance.

A few things I'd like to do here:

- Eliminate `bin/api-docs/packages.js`
   - The changes in this pull request to introduce some token-checking might allow us to apply similar filtering to unstaged README files, without needing to consult an explicit whitelist of files. Based on the conversation from https://github.com/WordPress/gutenberg/pull/18820#discussion_r353970649.
- Eliminate `docs/tool/update-data.js`
   - I'm thinking this can be collapsed into the existing behavior of `update-readmes.js`, where the file glob is updated to consider both package READMEs and this data documentation, and the data documentation can leverage the "file reference" syntax introduced here as a means to point to the actions/selectors files from where the documentation is to be extracted.

**Comment by aduth** [comment]
I've refreshed this pull request. As I mentioned in previous comments, I was able to absorb all of the functionality of data and package documentation to a single script and eliminate the hard-coded listing of packages.

Average run times are ~1.8s for a full run (down from ~6.3s). However, I expect the most common usage will only need to regenerate documentation for a single package (via pre-commit arguments), where the average run time is more in the range of ~0.7s.

**Comment by nosolosw** [comment]
I'm reviewing this (and pushed a path fix).

**Review by oandregal — CHANGES_REQUESTED** [review]
Tested this with [these testing steps](https://github.com/WordPress/gutenberg/pull/18840#issuecomment-561751133) and it fails to update the handbook file (with the steps described there, it'll be the `docs/designers-developers/developers/data/data-core.md` file).

**Comment by nosolosw** [comment]
In https://github.com/WordPress/gutenberg/pull/18840/commits/725a47efe73e5a0a85fe03679b74d424988980ad I've pushed a fix for [this bug](https://github.com/WordPress/gutenberg/pull/18840/#pullrequestreview-379343942). Still reviewing other packages.

**Comment by nosolosw** [comment]
Also found a JSDoc comment not adhering to the proper format so wasn't being picked up. Fixed at https://github.com/WordPress/gutenberg/pull/18840/commits/fd9129225f197b7027909880074a2befe45f4b51 This arguably can be in a different PR, adding it here for simplicity --- and because it added some fun times while debugging.

**Comment by nosolosw** [comment]
In https://github.com/WordPress/gutenberg/pull/18840/commits/8b4310f52bf9962bcde35f0b2b90d9be6b7179d0 and https://github.com/WordPress/gutenberg/pull/18840/commits/9ceac1f53484d5f1c16b5cb1c0cd335e40c7d1bf removed some leftovers.

File: bin/api-docs/update-api-docs.js [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+/**
+ * Returns the conventional package name of a given documentation file name.
+ *
+ * @param {string} file Documentation file name.
+ *
+ * @return {string} package name.
+ */
+function getDataDocumentationFilePackage( file ) {
```
Comment (nosolosw) [inline bin/api-docs/update-api-docs.js]: Following up on https://github.com/WordPress/gutenberg/pull/18840/files#r396461733

In which use cases this function is called?

File: packages/block-editor/src/store/actions.js [hunk 5]
```diff
@@ -788,7 +788,7 @@ export function updateBlockListSettings( clientId, settings ) {
 	};
 }
 
-/*
+/**
```
Comment (nosolosw) [inline packages/block-editor/src/store/actions.js]: See https://github.com/WordPress/gutenberg/pull/18840/#issuecomment-602586554

File: packages/docgen/README.md [hunk 6]
```diff
@@ -33,9 +33,9 @@ This command will generate a file named `entry-point-api.md` containing all the
 * **--output** `(String)`: Output file that will contain the API documentation.
 * **--to-section** `(String)`: Append generated documentation to this section in the Markdown output. To be used by the default Markdown formatter. Depends on `--output` and bypasses the custom `--formatter` passed, if any.
 * **--to-token**: Embed generated documentation within the start and end tokens in the Markdown output. To be used by the default Markdown formatter.Depends on `--output` and bypasses the custom `--formatter` passed, if any.
-  * Start token: `<!-- START TOKEN(Autogenerated API docs) -->`
+  * Start token: <code>&lt;!-- START TOKEN(Autogenerated API docs) --></code>
```
Comment (nosolosw) [inline packages/docgen/README.md]: Why is this necessary (and not in the item list just below this one)?

**Comment by nosolosw** [comment]
Have to take a break, but will come back to this today.

File: bin/api-docs/update-api-docs.js [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+	ROOT_DIR,
+	'docs/designers-developers/developers/data'
+);
+
+/**
+ * Pattern matching start token of a README file.
+ *
+ * @type {RegExp}
+ */
+const TOKEN_PATTERN = /<!-- START TOKEN\((.+?(?:\|(.+?))?)\) -->/g;
```
Comment (nosolosw) [inline bin/api-docs/update-api-docs.js]: Can we document what this RegExp is targeting? Perhaps adding two examples (one with matched path and other with default path) would be helpful.

**Comment by nosolosw** [comment]
@aduth I've done with testing and code review. I've also taken the liberty of pushing some code myself, hope that you don't mind! I wanted to accelerate this because it's such a good improvement.

I left a couple of comments. Especially I'd like to know more about [this](https://github.com/WordPress/gutenberg/pull/18840#discussion_r396461733) before approving, to make sure I didn't miss any use case in my testing.

File: bin/api-docs/update-api-docs.js [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+/**
+ * Given an absolute file path, returns the package name.
+ *
+ * @param {string} file Absolute path.
+ *
+ * @return {string} Package name.
+ */
+function getFilePackage( file ) {
+	return file.startsWith( PACKAGES_DIR )
```
[the remark this answers is not in the document]
Comment (aduth) [reply bin/api-docs/update-api-docs.js]: I think this was most relevant prior to 37adca35027a1ba3bf7d16ecaff6b65c6772ba94 , where previously the path included in the token would try to resolve from the package root, rather than from where the documentation resided.

If it's not referenced elsewhere anymore, it may be safe to remove this condition.

File: bin/api-docs/update-api-docs.js (hunk 4 above)
Comment (aduth) [reply bin/api-docs/update-api-docs.js]: Likely the same response from https://github.com/WordPress/gutenberg/pull/18840#discussion_r396771738 😅 I had a feeling that there was some further simplification I might have been able to do after 37adca35027a1ba3bf7d16ecaff6b65c6772ba94, but I clearly didn't follow the logic thoroughly enough to see where those simplifications could be made.

File: packages/docgen/README.md (hunk 6 above)
Comment (aduth) [reply packages/docgen/README.md]: I recall in initial implementations that I was seeing that docgen was trying to insert documentation here 😆 Without digging further, it's not clear exactly why it wouldn't be an issue for the other instance of token pairs.

File: bin/api-docs/update-api-docs.js (hunk 7 above)
Comment (aduth) [reply bin/api-docs/update-api-docs.js]: I [recently](https://github.com/aduth/dones/commit/012649c54cb37e041c2a124eb7649cf560c4c10b#diff-a73199f6b673a7742e2f8377e84da5d8R37-R43) became acquainted with [ES2018 named capture groups](https://2ality.com/2017/05/regexp-named-capture-groups.html). Maybe that could help make the pattern itself a little clearer to decipher? Alternatively, I've tried [other approaches at documenting capture groups](https://github.com/aduth/tannin/blob/cd1c7447843df7751c4abd1b92aee03fe56bfba4/packages/sprintf/index.js#L33-L38). Your suggestion of a couple simple examples is probably a good idea to include in any case.

**Comment by aduth** [comment]
> @aduth I'm done with testing and code review. I've also taken the liberty of pushing some code myself, hope that you don't mind! I wanted to accelerate this because it's such a good improvement.

I don't mind at all! Quite the opposite, in fact. I appreciate the assist.

I only did a cursory round of responses to your feedback, and I don't think I'll find the time today to dig into the details. I'm eager to keep it moving, though!

**Comment by aduth** [comment]
Regarding 4179376ee273dfe05962100c3957a82d11d888f7 : I think you may be correct in the current state of affairs that it's unnecessary. However, if I recall correctly, the reason I had implemented it this way initially was in anticipation of potential undesirable warnings / behavior, related to:

- https://thecodebarbarian.com/unhandled-promise-rejections-in-node.js.html
- https://nodejs.org/api/process.html#process_event_rejectionhandled

I'd need to dig into those resources again to fully understand the possible implications and whether we're "safe" without the `try` / `catch`, but just highlighting as the reason it was there in the first place.

File: bin/api-docs/update-api-docs.js (hunk 7 above)
Comment (nosolosw) [reply bin/api-docs/update-api-docs.js]: Added the two use cases this is trying to match at https://github.com/WordPress/gutenberg/pull/18840/commits/0475f24e8de08fece235547563b17c463f9c35a6 

Named capture groups look nice. Also, I applaud the effort in documenting the tannin regexp :clap:

File: bin/api-docs/update-api-docs.js (hunk 4 above)
Comment (nosolosw) [reply bin/api-docs/update-api-docs.js]: https://github.com/WordPress/gutenberg/pull/18840/commits/77dd061a875031d22bd87901c1461333e0a1c49e

File: bin/api-docs/update-api-docs.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (nosolosw) [reply bin/api-docs/update-api-docs.js]: https://github.com/WordPress/gutenberg/pull/18840/commits/77dd061a875031d22bd87901c1461333e0a1c49e

File: packages/docgen/README.md (hunk 6 above)
Comment (nosolosw) [reply packages/docgen/README.md]: Ah! I now see how this file is matched by the script: when you run the script without any files as input, it'll search the whole codebase for candidates. Made a few tweaks at https://github.com/WordPress/gutenberg/pull/18840/commits/4119f471d3b2ab222420eb67a9b74a5fa05fc3e1 for consistency.

**Review by oandregal — APPROVED** [review]
Very excited about the performance gains and simplification this brings in: in my testing, it takes 1/3 the time it took to validate and generate the files. Thanks for your work, Andrew!

**Comment by aduth** [comment]
@sirreal This may be an interesting case to consider for your work in #18942. The changes here seek to opt-in the `bin/` file to type-checking. In #18942, we're removing this configuration, effectively only allowing packages to be type-checked.

Can you imagine there to be any compromise? Should we just accept that these scripts won't be type-checked? Would individual top-of-file `// @ts-check` flags suffice?

File: docs/designers-developers/developers/data/data-core.md [hunk 9]
```diff
@@ -617,4 +616,5 @@ _Parameters_
 Action triggered to undo the last edit to
 an entity record, if any.
 
-<!-- END TOKEN(Autogenerated actions) -->
+
+<!-- END TOKEN(Autogenerated actions|../../../../packages/core-data/src/actions.js) -->
```
Comment (aduth) [inline docs/designers-developers/developers/data/data-core.md]: Full-transparency: I'm not entirely sure that the extra newline here would be expected. And, if I recall correctly, these were some of the symptoms we saw in the earlier parallelization efforts that ultimately led to their revert. But when I debugged the code flow of the documentation build script, it was always the case that no two `docgen` processes would run on the same document at the same time, which was the problem we had in the earlier implementation.

**Comment by aduth** [comment]
Rebased, since there were failures related to yesterday's issues (#21118).

It's end of day for me, so I'll plan to merge this tomorrow assuming everything is in order, and to allow time for the build to pass and potential responses to https://github.com/WordPress/gutenberg/pull/18840#issuecomment-604083497 and/or https://github.com/WordPress/gutenberg/pull/18840#discussion_r398166812. I'm not too worried if we just have to remove this from type-checking if it becomes a problem for #18942.

**Comment by sirreal** [comment]
> @sirreal This may be an interesting case to consider for your work in #18942.

Thanks for bringing this up.

> The changes here seek to opt-in the `bin/` file to type-checking. In #18942, we're removing this configuration, effectively only allowing packages to be type-checked.

I don't believe anything in #18942 will be compatible with the changes here. The latest iterations in #18942 continue to include "index" into all the typed projects (not necessarily packages). Scripts in `bin` could be typechecked as well, either as part of the main build or independently.

https://github.com/WordPress/gutenberg/blob/31ea1e753d2f37758ac778d6f757dbe887859142/tsconfig.json#L2-L13

File: docs/designers-developers/developers/data/data-core.md (hunk 9 above)
Comment (nosolosw) [reply docs/designers-developers/developers/data/data-core.md]: This is something that I've looked and tested specifically by modifying core-data's selectors and actions, whose changes should be reflected in the package's README and the handbook's data-core doc. It's working as expected: different files are processed asynchronously (good for performance), but different tokens within the same file are processed synchronously (required for docgen). So that's :+1: 

I don't have an answer for the extra line, though. It gives me confidence that it's consistent and doesn't depend on input order: note how the extra line is always added before the end of the second token for all files, while in the `master` implementation that extra line was always added at the end of the first token. In data-core.md the last item is the actions while in the README.md is the selectors. I've also tested for things like input variance (reversed input data and output is the same):

`node ./bin/api-docs/update-api-docs.js packages/core-data/src/selectors.js packages/core-data/src/actions.js`

and 

`node ./bin/api-docs/update-api-docs.js packages/core-data/src/actions.js packages/core-data/src/selectors.js`

produce the same results.

**Comment by aduth** [comment]
> I don't believe anything in #18942 will be compatible with the changes here.

Just to clarify: Do you mean "incompatible"? The remainder of your comment would seem to suggest it'll be possible to include.

**Comment by sirreal** [comment]
> > I don't believe anything in #18942 will be compatible with the changes here.
> 
> Just to clarify: Do you mean "incompatible"? The remainder of your comment would seem to suggest it'll be possible to include.

🤦‍♂ Completely _compatible_, _no incompatibilities_ 👍 

#18942 has been rebased and includes this in its type build/check setup.

**Comment by aduth** [comment]
> Regarding [4179376](https://github.com/WordPress/gutenberg/commit/4179376ee273dfe05962100c3957a82d11d888f7) : I think you may be correct in the current state of affairs that it's unnecessary. However, if I recall correctly, the reason I had implemented it this way initially was in anticipation of potential undesirable warnings / behavior, related to:
> 
> * https://thecodebarbarian.com/unhandled-promise-rejections-in-node.js.html
> * https://nodejs.org/api/process.html#process_event_rejectionhandled
> 
> I'd need to dig into those resources again to fully understand the possible implications and whether we're "safe" without the `try` / `catch`, but just highlighting as the reason it was there in the first place.

It turns out that it is necessary 😅 See https://github.com/WordPress/gutenberg/pull/21467#discussion_r410403049 . I'll be preparing a pull request shortly.
